### PR TITLE
fix(scaffolder): fix duplicate stream logs when switching tabs

### LINKED ISSUE #25989: 🐛 Bug Report: Scaffolder Logs Loading twice when I switch tabs
### 📜 Description

The same Scaffolder Logs loads again when I try to switch tabs.

### 👍 Expected behavior

The same logs shouldn't load twice

### 👎 Actual Behavior with Screenshots

https://github.com/user-attachments/assets/e2331b0f-66a9-40c3-add0-613bf4cf7a7c

### 👟 Reproduction steps

Run any scaffolder task, in the example I'm using the same `Run of example-nodejs-template` that comes with backstage installation. And then change tabs while the task is in progress. 

### 📃 Provide the context for the Bug.

The logs becomes repetitive and confusing. 

### 🖥️ Your Environment

```
OS:   Darwin 23.4.0 - darwin/arm64
node: v18.19.0
yarn: 1.22.19
cli:  0.26.11 (installed)
backstage:  1.29.0

Dependencies:
  @backstage/app-defaults                                          1.5.9
  @backstage/backend-app-api                                       0.8.0
  @backstage/backend-common                                        0.23.3
  @backstage/backend-defaults                                      0.4.1
  @backstage/backend-dev-utils                                     0.1.4
  @backstage/backend-openapi-utils                                 0.1.15
  @backstage/backend-plugin-api                                    0.7.0
  @backstage/backend-tasks                                         0.5.27
  @backstage/catalog-client                                        1.6.5
  @backstage/catalog-model                                         1.5.0
  @backstage/cli-common                                            0.1.14
  @backstage/cli-node                                              0.2.7
  @backstage/cli                                                   0.26.11
  @backstage/config-loader                                         1.8.1
  @backstage/config                                                1.2.0
  @backstage/core-app-api                                          1.14.1
  @backstage/core-compat-api                                       0.2.7
  @backstage/core-components                                       0.13.10, 0.14.9
  @backstage/core-plugin-api                                       1.9.3
  @backstage/e2e-test-utils                                        0.1.1
  @backstage/errors                                                1.2.4
  @backstage/eslint-plugin                                         0.1.8
  @backstage/frontend-plugin-api                                   0.6.7
  @backstage/integration-aws-node                                  0.1.12
  @backstage/integration-react                                     1.1.29
  @backstage/integration                                           1.13.0
  @backstage/plugin-api-docs                                       0.11.7
  @backstage/plugin-app-backend                                    0.3.71
  @backstage/plugin-app-node                                       0.1.22
  @backstage/plugin-auth-backend-module-atlassian-provider         0.2.3
  @backstage/plugin-auth-backend-module-aws-alb-provider           0.1.14
  @backstage/plugin-auth-backend-module-azure-easyauth-provider    0.1.5
  @backstage/plugin-auth-backend-module-bitbucket-provider         0.1.5
  @backstage/plugin-auth-backend-module-cloudflare-access-provider 0.1.5
  @backstage/plugin-auth-backend-module-gcp-iap-provider           0.2.17
  @backstage/plugin-auth-backend-module-github-provider            0.1.19
  @backstage/plugin-auth-backend-module-gitlab-provider            0.1.19
  @backstage/plugin-auth-backend-module-google-provider            0.1.19
  @backstage/plugin-auth-backend-module-guest-provider             0.1.8
  @backstage/plugin-auth-backend-module-microsoft-provider         0.1.17
  @backstage/plugin-auth-backend-module-oauth2-provider            0.2.3
  @backstage/plugin-auth-backend-module-oauth2-proxy-provider      0.1.15
  @backstage/plugin-auth-backend-module-oidc-provider              0.2.3
  @backstage/plugin-auth-backend-module-okta-provider              0.0.15
  @backstage/plugin-auth-backend-module-onelogin-provider          0.1.3
  @backstage/plugin-auth-backend                                   0.22.9
  @backstage/plugin-auth-node                                      0.4.17
  @backstage/plugin-auth-react                                     0.1.4
  @backstage/plugin-bitbucket-cloud-common                         0.2.21
  @backstage/plugin-catalog-backend-module-logs                    0.0.1
  @backstage/plugin-catalog-backend-module-scaffolder-entity-model 0.1.20
  @backstage/plugin-catalog-backend                                1.24.0
  @backstage/plugin-catalog-common                                 1.0.25
  @backstage/plugin-catalog-graph                                  0.4.7
  @backstage/plugin-catalog-import                                 0.12.1
  @backstage/plugin-catalog-node                                   1.12.4
  @backstage/plugin-catalog-react                                  1.12.2
  @backstage/plugin-catalog                                        1.21.1
  @backstage/plugin-events-node                                    0.3.8
  @backstage/plugin-org                                            0.6.27
  @backstage/plugin-permission-backend-module-allow-all-policy     0.1.19
  @backstage/plugin-permission-backend                             0.5.46
  @backstage/plugin-permission-common                              0.8.0
  @backstage/plugin-permission-node                                0.8.0
  @backstage/plugin-permission-react                               0.4.24
  @backstage/plugin-proxy-backend                                  0.5.3
  @backstage/plugin-scaffolder-backend-module-azure                0.1.14
  @backstage/plugin-scaffolder-backend-module-bitbucket-cloud      0.1.12
  @backstage/plugin-scaffolder-backend-module-bitbucket-server     0.1.12
  @backstage/plugin-scaffolder-backend-module-bitbucket            0.2.12
  @backstage/plugin-scaffolder-backend-module-gerrit               0.1.14
  @backstage/plugin-scaffolder-backend-module-gitea                0.1.12
  @backstage/plugin-scaffolder-backend-module-github               0.4.0
  @backstage/plugin-scaffolder-backend-module-gitlab               0.4.4
  @backstage/plugin-scaffolder-backend                             1.23.0
  @backstage/plugin-scaffolder-common                              1.5.4
  @backstage/plugin-scaffolder-node                                0.4.8
  @backstage/plugin-scaffolder-react                               1.10.0
  @backstage/plugin-scaffolder                                     1.23.0
  @backstage/plugin-search-backend-module-catalog                  0.1.28
  @backstage/plugin-search-backend-module-pg                       0.5.32
  @backstage/plugin-search-backend-module-techdocs                 0.1.27
  @backstage/plugin-search-backend-node                            1.2.27
  @backstage/plugin-search-backend                                 1.5.14
  @backstage/plugin-search-common                                  1.2.13
  @backstage/plugin-search-react                                   1.7.13
  @backstage/plugin-search                                         1.4.14
  @backstage/plugin-signals-react                                  0.0.4
  @backstage/plugin-techdocs-backend                               1.10.9
  @backstage/plugin-techdocs-module-addons-contrib                 1.1.12
  @backstage/plugin-techdocs-node                                  1.12.8
  @backstage/plugin-techdocs-react                                 1.2.6
  @backstage/plugin-techdocs                                       1.10.7
  @backstage/plugin-user-settings-common                           0.0.1
  @backstage/plugin-user-settings                                  0.8.10
  @backstage/release-manifests                                     0.0.11
  @backstage/test-utils                                            1.5.9
  @backstage/theme                                                 0.5.6
  @backstage/types                                                 1.1.1
  @backstage/version-bridge                                        1.0.8
```

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by carlerikjohan**
## Hey, I just made a Pull Request!

The scaffolder task event stream replayed all log entries when the browser tab
regained focus, causing logs to appear twice. The underlying SSE library
reconnected silently without a cursor, and the Observable never aborted the
connection on unsubscribe.

This fixes connection lifecycle management in `ScaffolderClient` and adds
cursor-based reconnection and visibility-change handling in the event stream
hook so that only new events are fetched on resume.

This would fix: https://github.com/backstage/backstage/issues/25989

## How it worked before

https://github.com/user-attachments/assets/fc45f433-7d79-4a92-a3d9-c44d6640378d

## How it works now

https://github.com/user-attachments/assets/b536d474-9a3a-48a4-b583-33a4051ecad4

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/scaffolder-react/src/hooks/useEventStream.ts [hunk 1]
```diff
@@ -194,108 +194,147 @@ export const useTaskEventStream = (taskId: string): TaskStream => {
...
           }
         },
       );
+    };
     void startStreamLogProcess();
+
+    const onVisibilityChange = () => {
+      if (didCancel || isStreamComplete) return;
+      if (document.hidden) {
+        cleanUpStream();
```
Comment (freben) [inline plugins/scaffolder-react/src/hooks/useEventStream.ts:323]: `cleanUpStream` clears the 500 ms batching interval without flushing `collectedLogEvents`. Since `lastSeenEventId` is advanced as soon as each event arrives, hiding the tab during that window discards the buffered logs, and the resumed request's `after` cursor ensures the backend will not replay them. Could we flush before visibility teardown, or only advance the cursor once the buffered events have been dispatched? Please add a fake-timer regression test covering log received -> hide before 500 ms -> show.

File: plugins/scaffolder-react/src/hooks/useEventStream.ts [hunk 2]
```diff
@@ -194,108 +194,147 @@ export const useTaskEventStream = (taskId: string): TaskStream => {
...
+            }/${maxRetries} retries)`;
+          }
+
+          retryTimer = setTimeout(() => {
+            retryTimer = undefined;
+            retryCount += 1;
+            reconnectStream();
+          }, 15000);
+
+          dispatch({ type: 'ERROR', data: error });
```
Comment (freben) [inline plugins/scaffolder-react/src/hooks/useEventStream.ts:289]: This sets `completed = true` and stores the transport error, but the retry now calls `reconnectStream` directly rather than dispatching `INIT`. Neither reconnection nor subsequent `LOGS` clears those fields, so the stream can recover and receive fresh events while the UI still displays an error and disables cancellation. Could we introduce a reconnect state transition that clears the transient transport error/completed state, with a test covering error -> retry -> new log?

File: plugins/scaffolder-common/src/ScaffolderClient.ts [hunk 3]
```diff
@@ -263,10 +266,10 @@ export class ScaffolderClient implements ScaffolderApi {
             }
           };
 
-          const ctrl = new AbortController();
           void fetchEventSource(url, {
             fetch: this.fetchApi.fetch,
             signal: ctrl.signal,
+            openWhenHidden: true,
```
Comment (freben) [inline plugins/scaffolder-common/src/ScaffolderClient.ts:272]: There is still an early-unsubscribe race here: teardown can abort `ctrl` while `getBaseUrl` is pending, after which this calls `fetchEventSource` with an already-aborted signal. This library only installs an abort listener; it does not check `signal.aborted`, so the already-fired abort is missed and an SSE request is still opened. Please check `ctrl.signal.aborted` or `subscriber.closed` before starting the request and cover unsubscribing before discovery resolves.

File: plugins/scaffolder-react/src/hooks/useEventStream.ts [hunk 4]
```diff
@@ -194,108 +194,147 @@ export const useTaskEventStream = (taskId: string): TaskStream => {
...
+      } else if (hasInitialized) {
+        reconnectStream();
+      } else {
+        void startStreamLogProcess();
+      }
+    };
+    document.addEventListener('visibilitychange', onVisibilityChange);
+
     return () => {
       if (!isTaskRecoverable) {
```
Comment (freben) [inline plugins/scaffolder-react/src/hooks/useEventStream.ts:333]: Recoverable tasks still bypass effect teardown entirely, leaving `didCancel` false and retaining the subscription, batching interval, or pending retry after unmount/dependency change. A later remount can therefore run a second stream alongside the detached one and recreate duplicate events. Recoverability should control stream completion semantics, not ownership of React resources; can we always mark the effect cancelled and call `cleanUpStream` here?

**Review by freben — CHANGES_REQUESTED** [review]
The overall direction is right, but the new visibility and reconnect lifecycle still has correctness gaps that can lose logs, leave the UI in a stale terminal state, or retain duplicate streams. The client-level tests are green, but these cases need hook-level lifecycle coverage before merging.

File: plugins/scaffolder-react/src/hooks/useEventStream.ts (hunk 1 above)
Comment (carlerikjohan) [reply plugins/scaffolder-react/src/hooks/useEventStream.ts:349]: Fixed. ```cleanUpStream``` now calls a ```currentEmitLogs``` ref to flush buffered logs before clearing the interval. Applied the same to the cancelled handler which had the same gap. Added a fake-timer test covering the hide-before-flush and cursor-on-resume paths.

File: plugins/scaffolder-react/src/hooks/useEventStream.ts (hunk 2 above)
Comment (carlerikjohan) [reply plugins/scaffolder-react/src/hooks/useEventStream.ts:314]: Added a ```RECONNECTING``` reducer action (clears error, resets completed) dispatched from ```reconnectStream```. Also added an ```isStreamComplete``` guard in the error handler to prevent an infinite 15s reconnect loop when the server closes the connection after a terminal event. recovered resets ```isStreamComplete``` so retry stays functional after recovery. Tests cover error-retry-clears-state and cancellation-stops-retry.

File: plugins/scaffolder-common/src/ScaffolderClient.ts (hunk 3 above)
Comment (carlerikjohan) [reply plugins/scaffolder-common/src/ScaffolderClient.ts:276]: Added ```if (ctrl.signal.aborted) { return; }``` before ```fetchEventSource```. Test defers the discovery promise, unsubscribes, resolves, asserts ```fetchEventSource``` was never called.

File: plugins/scaffolder-react/src/hooks/useEventStream.ts (hunk 4 above)
Comment (carlerikjohan) [reply plugins/scaffolder-react/src/hooks/useEventStream.ts]: Cleanup now always sets ```didCancel = true``` and calls ```cleanUpStream()```. Test verifies teardown on unmount for a recoverable task.

File: plugins/scaffolder-react/src/hooks/useEventStream.ts [hunk 5]
```diff
@@ -194,109 +201,154 @@ export const useTaskEventStream = (taskId: string): TaskStream => {
...
+              dispatch({ type: 'RECOVERED', data: event });
+              return undefined;
+            default:
+              throw new Error(`Unhandled event type ${event.type} in observer`);
+          }
+        },
+        error: error => {
+          cleanUpStream();
+
+          if (isStreamComplete || generation !== startGeneration || didCancel) {
```
Comment (freben) [inline plugins/scaffolder-react/src/hooks/useEventStream.ts:296]: `isStreamComplete` is set for both `completion` and `cancelled`, but recoverable-task streams deliberately remain open so that they can later receive `recovered`. If the connection closes while the completed/cancelled task is waiting for the user to retry it, this guard suppresses reconnection permanently, so the subsequent `recovered` event and new logs are never observed without a page refresh. Could we only treat these events as stream-terminal when `isTaskRecoverable` is false, and add a regression test covering completion -> transport error -> retry/recovered?

File: plugins/scaffolder-react/src/hooks/useEventStream.ts [hunk 6]
```diff
@@ -194,109 +201,154 @@ export const useTaskEventStream = (taskId: string): TaskStream => {
...
-          subscription.unsubscribe();
-        }
-        if (logPusher) {
-          clearInterval(logPusher);
-        }
+
+    const onVisibilityChange = () => {
+      if (didCancel || isStreamComplete) return;
+      if (document.hidden) {
+        cleanUpStream();
```
Comment (freben) [inline plugins/scaffolder-react/src/hooks/useEventStream.ts:349]: If the tab becomes hidden while the initial `getTask` call is pending, this cleanup does not invalidate that request generation. When `getTask` resolves, it still calls `subscribeToStream` and opens an SSE connection while hidden; showing the tab then tears it down and opens another one. Could we invalidate/increment the generation on hide and test hide-before-`getTask`-resolves?

**Review by freben — CHANGES_REQUESTED** [review]
The previous lifecycle concerns are resolved, but two edge cases remain. The recoverable-task disconnect is blocking because it can prevent the UI from observing a later retry until the page is refreshed.

File: plugins/scaffolder-react/src/hooks/useEventStream.ts (hunk 5 above)
Comment (carlerikjohan) [reply plugins/scaffolder-react/src/hooks/useEventStream.ts:296]: Fixed. ```isStreamComplete``` is now only set for non-recoverable tasks. Recoverable tasks keep the reconnection path open so recovered events are still observed after a transport error. Added a regression test covering completion -> transport error -> retry for a recoverable task.

File: plugins/scaffolder-react/src/hooks/useEventStream.ts (hunk 6 above)
Comment (carlerikjohan) [reply plugins/scaffolder-react/src/hooks/useEventStream.ts:349]: Fixed. The visibility-hide handler now increments ```startGeneration``` before cleanup, so a pending ```getTask``` resolution sees a stale generation and bails out. When the tab is shown again, ```startStreamLogProcess``` is called fresh. Added a test covering hide-before-getTask-resolves.

File: plugins/scaffolder-common/src/ScaffolderClient.ts [hunk 7]
```diff
@@ -279,15 +286,22 @@ export class ScaffolderClient implements ScaffolderApi {
               }
               processEvent(e);
             },
+            onclose() {
+              subscriber.error(new Error('SSE connection closed unexpectedly'));
+            },
             onerror(err) {
+              ctrl.abort();
               subscriber.error(err);
+              throw err;
```
Comment (freben) [inline plugins/scaffolder-common/src/ScaffolderClient.ts:295]: Throwing here is the documented way to stop `fetchEventSource` from retrying, but it also rejects the promise returned by `fetchEventSource`. Since that promise is discarded with `void`, every transport error will now surface as an unhandled promise rejection in addition to the observable error. The test currently hides this by catching the throw inside the mock implementation. Please consume the returned rejection (or propagate the error from its rejection handler) so this expected reconnect path does not produce an unhandled rejection.

**Review by freben — CHANGES_REQUESTED** [review]
The previous lifecycle concerns are resolved in the latest update. One error-handling issue remains: the fatal `fetchEventSource` error is now also left as an unhandled rejected promise.

File: plugins/scaffolder-common/src/ScaffolderClient.ts (hunk 7 above)
Comment (carlerikjohan) [reply plugins/scaffolder-common/src/ScaffolderClient.ts:295]: Fixed. Added .catch(() => {}) on the fetchEventSource promise with a comment explaining why. The error is already forwarded via subscriber.error before the throw, so the catch just consumes the expected rejection.

File: plugins/scaffolder-react/src/hooks/useEventStream.ts [hunk 8]
```diff
@@ -194,109 +201,159 @@ export const useTaskEventStream = (taskId: string): TaskStream => {
...
-        didCancel = true;
-        if (subscription) {
-          subscription.unsubscribe();
-        }
-        if (logPusher) {
-          clearInterval(logPusher);
-        }
+
+    const onVisibilityChange = () => {
+      if (didCancel || isStreamComplete) return;
```
Comment (freben) [inline plugins/scaffolder-react/src/hooks/useEventStream.ts:346]: For a non-recoverable task, receiving `cancelled` sets `isStreamComplete`, but neither the client nor the backend closes the SSE stream on cancellation. This early return then prevents the visibility handler from cleaning it up, and `openWhenHidden: true` keeps it alive in the background—potentially indefinitely if the cancelled action is hung and never emits its later completion. Please terminate the subscription when a non-recoverable cancellation is received, with a test asserting teardown/abort rather than only asserting that no retry occurs.

**Review by freben — CHANGES_REQUESTED** [review]
The previous lifecycle and error-handling concerns are resolved, and CI is green. One cancellation teardown issue remains: a non-recoverable cancelled task can retain its SSE connection, including while the tab is hidden.

File: plugins/scaffolder-react/src/hooks/useEventStream.ts (hunk 8 above)
Comment (carlerikjohan) [reply plugins/scaffolder-react/src/hooks/useEventStream.ts:346]: Fixed. cleanUpStream() is now called after dispatching CANCELLED or COMPLETED when the stream is terminal (non-recoverable). Replaced the existing cancellation test to assert that unsubscribe is called, and added an equivalent test for completion teardown.

**Review by freben — APPROVED** [review]

**Review by benjdlambert — APPROVED** [review]
Nice stuff! and we killed a few TODOS!
